File: lib/kwartz/scanner.rb
###
### copyright(c) 2005 kuwata-lab all rights reserved
###
### $Id$
### $Release$
###

require 'kwartz/exception'
require 'kwartz/utility'

module Kwartz

   class ScanError < BaseError
      def initialize(errmsg, linenum, filename)
         super(errmsg, linenum, filename)
      end
   end


   module CharType

      ## return true if ch is white char
      def is_white?(ch)
         case ch
         when ?\ , ?\n , ?\r , ?\t
            return true
         end
         return false
      end

      ## return true if ch is alphabet
      def is_alpha?(ch)
         return ?a <= ch && ch <= ?z || ?A <= ch && ch <= ?Z
      end

      ## return true if ch is numeric
      def is_digit?(ch)
         return ?0 <= ch && ch <= ?9
      end

      ## return true if ch is alphabet, number or underbar
      def is_word?(ch)
         return ?a <= ch && ch <= ?z || ?A <= ch && ch <= ?Z || ?0 <= ch && ch <= ?9 || ch == ?_
      end

   end


   ##
   ## ex.
   ##  input = ARGF.read()
   ##  properties = {}
   ##  scanner = Scanner.new(input, properties)
   ##  while scanner.scan()
   ##     if scanner.token().is_a?(String)
   ##        puts scanner.token()
   ##     elsif scanner.token().is_a?(Symbol)
   ##        puts scanner.token(), "(value=#{scanner.value()})"
   ##     end
   ##  end
   ##
   class Scanner
      include CharType

      def initialize(input, properties={})
         reset(input, 0)
         @properties = properties
         @filename = @properties[:filename]
         @keywords = @@keywords
      end

      def reset(input, base_linenum=0)
         @input = input
         @base_linenum = base_linenum
         @index = 0
         @char = @input[@index]
         @linenum = 1
         @token = nil
         @value = nil
      end

      attr_reader :input, :filename, :token, :value
      
      def linenum
         #return @linenum + @base_linenum
         n = @linenum + @base_linenum
         return @char == ?\n ? n - 1 : n
      end

      def getchar()
         @index += 1
         @char = @input[@index]
         @linenum += 1 if @char == ?\n
         return @char
      end

      @@keywords = {
         'expand'   =>   :expand,
         'element'  =>   :element,
         'macro'    =>   :macro,

         'print'    =>   :print,
         'while'    =>   :while,
         'foreach'  =>   :foreach,
         'for'      =>   :for,
         'in'       =>   :in,
         'if'       =>   :if,
         'else'     =>   :else,
         'elseif'   =>   :elseif,
         #'require'  =>   :require,

         'true'	    =>   :true,
         'false'    =>   :false,
         'null'	    =>   :null,
         'empty'    =>   :empty,
      }


      def scan()
         ## skip white spaces
         ch = @char
         ch = getchar() while ch && is_white?(ch)

         ## EOF
         if ch == nil
            return @token = nil
         end

         ## line comment or region comment
         if ch == ?/
            ch = getchar()
            if ch == ?/			## line comment
               ch = getchar()
               ch = getchar() while ch != nil && ch != ?\n
               getchar()
               return scan()
            elsif ch == ?*		## region comment
               while ch = getchar()
                  break if ch == ?* && (ch = getchar()) == ?/
               end
               unless ch != nil
                  msg = "region comment is not closed."
                  raise ScanError.new(msg, linenum(), filename())
               end
               getchar()
               return scan()
            elsif ch == ?=
               getchar()
               return @token = '/='	## /=
            else
               return @token = '/'	## /
            end
         end

         ## :  ::  :::
         if ch == ?:
            ch = getchar()
            return @token = ':' if ch != ?:
            ch = getchar()
            return @token = '::' if ch != ?:
            s = ''
            s << ch.chr while (ch = getchar()) != nil && ch != ?\n
            @value = s
            #return @token = ':::'
            return @token = :rawcode
         end

         ## name
         if is_alpha?(ch) || ch == ?_
            s = ch.chr
            while (ch = getchar()) != nil && is_word?(ch)
               s << ch.chr
            end
            @value = s
            w = @keywords[@value]
            @token = w ? w : :name
            return @token
         end

         ## numeric
         if is_digit?(ch)
            s = ch.chr
            while (ch = getchar()) != nil && is_digit?(ch)
               s << ch.chr
            end
            if ch == ?.
               s << ch.chr
               while (ch = getchar()) != nil && is_digit?(ch)
                  s << ch.chr
               end
            end
            @value = s
            return @token = :numeric
         end

         ## .  .=  .+  .+=
         if ch == ?.
            ch = getchar()
            if ch == ?=
               getchar()
               return @token = '.='
            end
            if ch == ?+
               if (ch = getchar()) == ?=
                  getchar()
                  return @token = '.+='
               else
                  return @token = '.+'
               end
            end
            return @token = '.'
         end

         ## quoted string
         if ch == ?'					#'
            s = ""
            while (ch = getchar()) != nil && ch != ?'	#'
               if ch == ?\\
                  ch = getchar()
                  break if ch == nil
                  case ch
                  when ?\\ ;  s << '\\'
                  when ?'  ;  s << "'"				#"
                  else
                     s << '\\' << ch.chr
                  end
               else
                  s << ch.chr
               end
            end
            if ch == nil
               msg = 'string "\'" is not closed.'
               raise ScanError.new(msg, linenum(), filename())
            end
            getchar()
            @value = s
            #print "*** debug: @value=#{@value.inspect}\n"
            #print "*** debug: dump_str(@value)=#{Kwartz::Util::dump_str(@value)}\n"
            return @token = :string
         end

         ## double-quoted string
         if ch == ?"					#"
            s = ""
            while (ch = getchar()) != nil && ch != ?"	#"
               if ch == ?\\
                  if (ch = getchar()) == nil
                     break
                  end
                  case ch
                  when ?\\ ;  s << '\\'
                  when ?"  ;  s << '"'				#'
                  when ?n  ;  s << "\n"
                  when ?t  ;  s << "\t"
                  when ?r  ;  s << "\r"
                  else
                     s << '\\' << ch.chr
                  end
               else
                  s << ch.chr
               end
            end
            if ch == nil
               msg = "string '\"' is not closed."
               raise ScanError.new(msg, linenum(), filename())
            end
            getchar()
            @value = s
            return @token = :string
         end

         ## +  -  *  /  %  ^
         if ch == ?+ || ch == ?- || ch == ?* || ch == ?/ || ch == ?% || ch == ?^
            if (ch2 = getchar()) != nil && ch2 == ?=
               getchar()
               return @token = ch.chr + '='	# += -= *= /= %= ^=
            else
               return @token = ch.chr		# +  -	*  /  %	 ^
            end
         end

         ## <  <=  <%  <?
         if ch == ?<
            ch2 = getchar()
            if ch2 == ?=
               getchar()
               @token = '<='
            elsif ch2 == ?% || ch2 == ??
               #@value = '<' + ch2.chr + getrest()
               #@token = :rawcode
               s = ''
               ch = getchar()
               @token = ch == ?= ? :rawexpr : :rawcode
               ch = getchar() if ch == ?=
               while true
                  unless ch
                     msg = "`<#{ch2.chr}#{@token == :rawexpr ? '=' : ''}' is not closed."
                     raise ScanError.new(msg, linenum(), @filename)
                  end
                  if ch != ch2
                     s << ch.chr
                  elsif (ch = getchar()) == ?>
                     getchar()
                     break
                  else
                     s << ch2.chr
                     s << ch.chr
                  end
                  ch = getchar()
               end
               @value = s
            else
               @token = '<'
            end
            return @token
         end

         ## =  !  >  ==  !=  >=
         if ch == ?= || ch == ?! || ch == ?< || ch == ?>
            if (ch2 = getchar()) != nil && ch2 == ?=
               getchar()
               return @token = ch.chr + '='	# == != <= >=
            else
               return @token = ch.chr		# =  !	<  >
            end
         end

         ## &&  ||
         if ch == ?& || ch == ?|
            if (ch2 = getchar()) != nil && ch2 == ch
               getchar()
               return @token = ch.chr + ch2.chr		## && ||
            else
               msg = "'#{ch.chr}': invalid char."
               raise ScanError.new(msg, linenum(), filename())
            end
         end

         ## @stag, @cont, @etag, @element(marking)
         if ch == ?@
            s = ''
            while (ch = getchar()) != nil && is_word?(ch)
               s << ch.chr
            end
            if s.empty?
               msg = "'@' requires macro name."
               raise ScanError.new(msg, linenum(), filename())
            end
            @value = s
            return @token = '@'
         end

         ##  [:  [
         if ch == ?[
            if (ch = getchar()) != nil && ch == ?:
               getchar()
               return @token = '[:'
            else
               return @token = '['
            end
         end

         ##  #  ]  ,  (  )  ?  {  }  ;
         case ch
         when ?# , ?] , ?, , ?(, ?), ??, ?{, ?}, ?;
            getchar()
            return @token = ch.chr
         end

         ## invalid char
         msg = "'#{ch.chr}': invalid char."
         raise ScanError.new(msg, linenum(), filename())
      end


      def scan_all()
         s = ''
         scan()		# don't change to `while scan()' ! 
         while token() != nil
            if token().is_a?(String)		# string
               s << token()
               case token()
               when '@', ':::'
                  s << value()
               when '#element', '#macro'
                  s << ' ' << value()
               end
            else				# symbol
               case token()
               when :string
                  s << Kwartz::Util::dump_str(value())
               when :numeric
                  s << value()
               when :name
                  s << value()
               when :true, :false, :null, :nil, :empty
                  s << ':' << token().id2name
               when :rawcode
                  s << "<%#{value()}%>"
               when :rawexpr
                  s << "<%=#{value()}%>"
               when :rubycode
                  s << ':rubycode' << value()
               else
                  s << ':' << token().id2name
               end
            end
            s << "\n"
            scan()
         end
         return s
      end

   end	# end of class Scanner


end   # end of module Kwartz

if __FILE__ == $0
   input = ARGF.read()
   scanner = Kwartz::Scanner.new(input)
   print scanner.scan_all()
end

File: src/main/java/kwartz/Scanner.java
/*
 * $Rev$
 * $Release$
 * $Copyright$
 */
package kwartz;

import java.util.Map;
import java.util.HashMap;

public class Scanner {

	private String _filename;
	private String _input;
	private int    _input_length;
	private int    _linenum = 1;
	private int    _column = 0;
	private int    _index = -1;
	private int    _ch;
	private String _token_value;
	private int    _start_linenum;
	private int    _start_column;
	private boolean _is_ruleset = false;
	private LexicalException _error;
	private StringBuffer _sbuf = new StringBuffer();
	
	
	public String getTokenValue() {
		return _token_value;
	}
	
	public LexicalException getError() {
		return _error;
	}
	
	public int getLinenum() {
		return _linenum;
	}
	
	public int getColumn() {
		return _column;
	}
	
	public int getStartLinenum() {
		return _start_linenum;
	}
	
	public int getStartColumn() {
		return _start_column;
	}
	
	public void setRulesetMode(boolean is_ruleset) {
		_is_ruleset = is_ruleset;
	}

	
	public Scanner(String input) {
		this(input, null);
	}


	public Scanner(String input, String filename) {
		_filename = filename;
		_input = input == null ? "" : input;
		_input_length = _input.length();
		// init
		getChar();
	}


//	public int getChar() {
//		_index++;
//		if (_index >= _input_length)
//			return _ch = 0;
//		_ch = _input.charAt(_index);
//		if (_ch == '\n') {
//			_linenum++;
//			_column = 0;
//		}
//		else {
//			_column++;
//		}
//		return _ch;
//	}
	public int getChar() {
		_index++;
		if (_index >= _input_length)
			return _ch = 0;
		if (_ch == '\n') {
			_linenum++;
			_column = 1;
		}
		else {
			_column++;
		}
		_ch = _input.charAt(_index);
		return _ch;
	}
	
	
	public static boolean isAlphabet(int ch) {
		return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z';
	}
	
	
	public static boolean isDigit(int ch) {
		//return Character.isDigit(ch);
		return '0' <= ch && ch <= '9';
	}
	
	
	public static boolean isWhite(int ch) {
		return Character.isWhitespace(ch);
	}


	protected LexicalException error(String message, int linenum, int column) {
		return new LexicalException(message, _filename, linenum, column);
	}
	
	
	public int scanIdentifier() {
		int ch = _ch;
		if (!isAlphabet(ch) && ch != '_') {
			return 0;
		}
		StringBuffer sbuf = _sbuf;
		sbuf.setLength(0);
		sbuf.append((char)ch);
		while (isAlphabet(ch = getChar()) || isDigit(ch) || ch == '_') {
			sbuf.append((char)ch);
		}
		_token_value = sbuf.toString();
		return Token.IDENT;
	}


	/*
	public String scanInteger() {
		int ch = _ch;
		if (!isDigit(ch)) {
			return null;
		}
		StringBuffer sbuf = _sbuf;
		sbuf.setLength(0);
		sbuf.append((char)ch);
		while (isDigit(ch = getChar())) {
			sbuf.append((char)ch);
		}
		return sbuf.toString();
	}
	*/
	
	
	public int scanNumber() {
		int ch = _ch;
		if (!isDigit(ch)) {
			return 0;
		}
		StringBuffer sbuf = _sbuf;
		sbuf.setLength(0);
		sbuf.append((char)ch);
		while (isDigit(ch = getChar())) {
			sbuf.append((char)ch);
		}
		if (ch != '.') {
			_token_value = sbuf.toString();
			return Token.INTEGER;
		}
		sbuf.append((char)ch);
		while (isDigit(ch = getChar())) {
			sbuf.append((char)ch);
		}
		_token_value = sbuf.toString();
		return Token.FLOAT;
	}
	
	
	public int scanString() {
		if (_ch == '\'') {
			return scanSingleQuotedString();
		}
		else if (_ch == '"') {
			return scanDoubleQuotedString();
		}
		assert false;  /* unreachable */
		return 0;
	}
	
	
	public int scanSingleQuotedString() {
		int ch = _ch;
		if (ch != '\'')
			return 0;
		int start_linenum = _linenum;
		int start_column  = _column;
		StringBuffer sbuf = _sbuf;
		sbuf.setLength(0);
		while ((ch = getChar()) != '\'' && ch != 0) {
			if (ch == '\\') {
				ch = getChar();
				if (ch == 0)
					break;
				if (ch != '\\' && ch != '\'')
					sbuf.append('\\');
			}
			sbuf.append((char)ch);
		}
		if (ch == 0) {
			_error = error("'\\'': string is not closed.", start_linenum, start_column);
			return Token.ERROR;
		}
		assert ch == '\'';
		getChar();
		_token_value = sbuf.toString();
		return Token.STRING;
	}
	
	
	public int scanDoubleQuotedString() {
		int ch = _ch;
		if (ch != '"')
			return 0;
		int start_linenum = _linenum;
		int start_column = _column;
		StringBuffer sbuf = _sbuf;
		sbuf.setLength(0);
		while ((ch = getChar()) != '"' && ch != 0) {
			if (ch == '\\') {
				ch = getChar();
				switch (ch) {
				case 'n':  sbuf.append('\n');  break;
				case 'b':  sbuf.append('\b');  break;
				case 'r':  sbuf.append('\r');  break;
				case 't':  sbuf.append('\t');  break;
				case 'f':  sbuf.append('\f');  break;
				case '"':  sbuf.append('"');   break;
				case '\\': sbuf.append('\\');  break;
				default:   sbuf.append('\\');  sbuf.append((char)ch);
				}
			}
			else {
				sbuf.append((char)ch);
			}
		}
		if (ch == 0) {
			_error = error("'\"': string is not closed.", start_linenum, start_column);
			return Token.ERROR;
		}
		assert ch == '"';
		getChar();
		_token_value = sbuf.toString();
		return Token.STRING;
	}

	
	public int scan() {
		//return _token_id = _scan();
		return _scan();
	}
	
	
	private int _scan() {
		int ch = _ch;
		while (isWhite(ch)) {
			ch = getChar();
		}
		if (ch == 0)
			return 0;
		_start_linenum = _linenum;
		_start_column  = _column;

		/// '/', '/=', '/*', '//'
		if (ch == '/') {
			ch = getChar();
			if (ch == '*') {        /// region comment
				int start_linenum = _linenum;
				int start_column = _column;
				while (true) {
					while ((ch = getChar()) != '*' && ch != 0)
						;
					if (ch == 0) {
						_error = error("comment is not closed.", start_linenum, start_column);
						return Token.ERROR;
					}
					if ((ch = getChar()) != '/')  /// comment is closed
						continue;
					getChar();
					return _scan();  // call recursively
				}
			}
			else if (ch == '/') {   /// line comment
				while ((ch = getChar()) != '\n' && ch != 0)
					;
				if (ch == 0)
					return 0;
				return _scan();  // call recursively
			}
			else if (_is_ruleset) {
				return _invalidChar('/');
			}
			else if (ch == '=') {
				ch = getChar();
				return Token.SLASH_EQ;   // "/="
			}
			else {
				return '/';
			}
		}

		/// ruleset mode
		if (_is_ruleset) {
			return _scanForRuleset(ch);
		}
		
		/// keyword or identifier
		if (isAlphabet(ch) || ch == '_') {
			scanIdentifier();
			Integer obj = (Integer)__keywords.get(_token_value);
			return obj != null ? obj.intValue() : Token.IDENT;
		}
		
		/// number
		if (isDigit(ch)) {
			int start_linenum = _linenum;
			int start_column = _column;
			int token = scanNumber();
			if (isAlphabet(_ch) || _ch == '_') {
				String numstr = _token_value;
				scanIdentifier();
				_error = error("'" + numstr + _token_value + "': invalid word.", start_linenum, start_column);
				return Token.ERROR;
			}
			return token;
		}
		
		/// string
		if (ch == '"' || ch == '\'') {
			return scanString();
		}
		
		
		/// '<', '<=', '<%', '<%='
		if (ch == '<') {
			ch = getChar();
			if (ch == '=') {
				getChar();
				return Token.LE;  // "<="
			}
			else if (ch == '%') {
				int start_linenum = _linenum;
				int start_column = _column;
				ch = getChar();
				boolean is_expr = false;
				String start_pat = "<%";
				if (ch == '=') {
					ch = getChar();
					is_expr = true;
					start_pat = "<%=";
				}
				StringBuffer sbuf = _sbuf;
				sbuf.setLength(0);
				while (true) {
					while (ch != '%' && ch != 0) {
						sbuf.append((char)ch);
						ch = getChar();
					}
					if (ch == 0) {
						_error = error("'" + start_pat + "': native code is not closed.", start_linenum, start_column);
						return Token.ERROR;
					}
					ch = getChar();
					if (ch == '>')
						break;
					sbuf.append('%').append(ch);
					ch = getChar();
				}
				_token_value = sbuf.toString();
				return is_expr ? Token.NATIVE_EXPR : Token.NATIVE_STMT;
			}
			else {
				return '<';
			}
		}

		/// '+', '-', '*', '%', '=', '!', '>'
		if (   ch == '+' || ch == '-' || ch == '*' || ch == '%'
			|| ch == '=' || ch == '!' || ch == '>') {
			if (getChar() != '=')
				return ch;
			getChar();
			switch (ch) {
			case '+':  return Token.PLUS_EQ;       // '+='
			case '-':  return Token.MINUS_EQ;      // '-='
			case '*':  return Token.STAR_EQ;       // '*='
			case '/':  return Token.SLASH_EQ;      // '/='
			case '%':  return Token.PERCENT_EQ;    // '%='
			case '=':  return Token.EQ;            // '=='
			case '!':  return Token.NE;            // '!='
			case '>':  return Token.GE;            // '>='
			default:   assert false;
			}
		}

		/// '&&', '||'
		if (ch == '&' || ch == '|') {
			int ch2 = getChar();
			int linenum = _linenum;
			int column = _column;
			if (ch != ch2) {
				_error = error("'" + ch + "': invalid character.", linenum, column);
				return Token.ERROR;
			}
			ch2 = getChar();
			if (ch2 != '=')
				return ch == '&' ? Token.AND : Token.OR;  // "&&" : "||"
			getChar();
			return ch == '&' ? Token.AND_EQ : Token.OR_EQ;  // "&&=" : "||="
		}

		/// '{', '}', '(', ')', ']', ';', ',', '?', ':'
		if (   ch == '{' || ch == '}' || ch == '(' || ch == ')' || ch == ']'
			|| ch == ';' || ch == ',' || ch == '?' || ch == ':') {
			getChar();
			return ch;
		}
		
		/// '.', '.+', '.+='
		if (ch == '.') {
			if (getChar() != '+')
				return '.';
			if (getChar() != '=')
				return Token.CONCAT;  // '.+'
			getChar();
			return Token.CONCAT_EQ;   // '.+='
		}
		
		/// '[:', '['
		if (ch == '[') {
			if (getChar() != ':')
				return '[';  // '[';
			getChar();
			return Token.INDEX2;     // '[:'
		}
		
		//_token_value = Character.toString((char)ch);
		//return ":ERROR";
		return _invalidChar(ch);
	}
	
	
	private int _invalidChar(int ch) {
		_token_value = Character.toString((char)ch);
		_error = error("'" + (char)ch + "': invalid character.", _start_linenum, _start_column);
		return Token.ERROR;   // or ch ?
	}
	
	
	private int _scanForRuleset(int ch) {
		if (ch == '#' || ch == '.') {
			int ch2 = getChar();
			if (! isAlphabet(ch2) && ch2 != '_')
				return _invalidChar(ch);
			scanIdentifier();
			_token_value = Character.toString((char)ch) + _token_value;
			return Token.SELECTOR;
		}
		if (ch == '{' || ch == '}' || ch == ':' || ch == ';' || ch == ',') {
			getChar();
			return ch;
		}
		if (ch == '@') {
			int ch2 = getChar();
			if (! isAlphabet(ch2) && ch2 != '_')
				return _invalidChar(ch);
			scanIdentifier();
			return Token.COMMAND;
		}
		if (ch == '"' || ch == '\'') {
			return scanString();
		}
		if (isAlphabet(ch) || ch == '_') {
			scanIdentifier();
			String word = _token_value;
			Integer intobj = (Integer)__properties.get(word);
			return intobj != null ? intobj.intValue() : Token.IDENT; 
		}
		return _invalidChar(ch);
	}
	
	
	private static Map __keywords;
	private static Map __properties;
	static {
		__keywords = new HashMap();
		__keywords.put("if",       new Integer(Token.IF));
		__keywords.put("else",     new Integer(Token.ELSE));
		__keywords.put("elseif",   new Integer(Token.ELSEIF));
		__keywords.put("foreach",  new Integer(Token.FOREACH));
		__keywords.put("while",    new Integer(Token.WHILE));
		__keywords.put("print",    new Integer(Token.PRINT));
		__keywords.put("break",    new Integer(Token.BREAK));
		__keywords.put("continue", new Integer(Token.CONTINUE));
		__keywords.put("true",     new Integer(Token.TRUE));
		__keywords.put("false",    new Integer(Token.FALSE));
		__keywords.put("null",     new Integer(Token.NULL));
		__keywords.put("_stag",    new Integer(Token.STAG));
		__keywords.put("_cont",    new Integer(Token.CONT));
		__keywords.put("_etag",    new Integer(Token.ETAG));
		__keywords.put("_elem",    new Integer(Token.ELEM));
		__keywords.put("_element", new Integer(Token.ELEMENT));
		__keywords.put("_content", new Integer(Token.CONTENT));
		
		__properties = new HashMap();
		__properties.put("stag",    new Integer(Token.P_STAG));
		__properties.put("Stag",    new Integer(Token.P_STAG));
		__properties.put("STAG",    new Integer(Token.P_STAG));
		__properties.put("etag",    new Integer(Token.P_ETAG));
		__properties.put("Etag",    new Integer(Token.P_ETAG));
		__properties.put("ETAG",    new Integer(Token.P_ETAG));
		__properties.put("cont",    new Integer(Token.P_CONT));
		__properties.put("Cont",    new Integer(Token.P_CONT));
		__properties.put("CONT",    new Integer(Token.P_CONT));
		__properties.put("elem",    new Integer(Token.P_ELEM));
		__properties.put("Elem",    new Integer(Token.P_ELEM));
		__properties.put("ELEM",    new Integer(Token.P_ELEM));
		__properties.put("value",   new Integer(Token.P_VALUE));
		__properties.put("Value",   new Integer(Token.P_VALUE));
		__properties.put("VALUE",   new Integer(Token.P_VALUE));
		__properties.put("attrs",   new Integer(Token.P_ATTRS));
		__properties.put("Attrs",   new Integer(Token.P_ATTRS));
		__properties.put("ATTRS",   new Integer(Token.P_ATTRS));
		__properties.put("append",  new Integer(Token.P_APPEND));
		__properties.put("Append",  new Integer(Token.P_APPEND));
		__properties.put("APPEND",  new Integer(Token.P_APPEND));
		__properties.put("remove",  new Integer(Token.P_REMOVE));
		__properties.put("tagname", new Integer(Token.P_TAGNAME));
		__properties.put("logic",   new Integer(Token.P_LOGIC));
		__properties.put("begin",   new Integer(Token.P_BEGIN));
		__properties.put("end",     new Integer(Token.P_END));
		__properties.put("before",  new Integer(Token.P_BEFORE));
		__properties.put("after",   new Integer(Token.P_AFTER));
		__properties.put("global",  new Integer(Token.P_GLOBAL));
	}

	
	public static void main(String[] args) throws Exception {
		String input;
		input = "if (x > 0) { print(x); }";
		input = "while (x && y || z) { x + y * z / 3 .+ \"foo\"; }";
		
		Scanner scanner = new Scanner(input);
		int token;
		while ((token = scanner.scan()) != 0) {
			System.out.println("token="+TokenHelper.tokenSymbol(token)+"("+token+"), value="+scanner.getTokenValue());
		}
	}


}
